### PR TITLE
Refactor fewshot examples to use task generics instead of `**kwargs`

### PR DESCRIPTION
**Pull request by rmitsch**
## Description

- Refactor fewshot examples to use task generics instead of `**kwargs`.
- Remove Python 3.6 support.

### Corresponding documentation PR

\-

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran all tests in `tests` and `usage_examples/tests`, and all new and existing tests passed. This includes
  - all external tests (i. e. `pytest` ran with `--external`)
  - all tests requiring a GPU (i. e. `pytest` ran with `--gpu`)
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy_llm/tasks/rel/util.py [hunk 1]
```diff
@@ -24,26 +23,9 @@ class EntityItem(BaseModel):
     label: str
 
 
-class RELExample(FewshotExample):
+class BaseRELExample(FewshotExample[TaskContraT], abc.ABC, Generic[TaskContraT]):
```
Comment (svlandeg) [inline spacy_llm/tasks/rel/util.py:26]: Why the renaming?

File: spacy_llm/tasks/span/examples.py [hunk 2]
```diff
@@ -1,13 +1,17 @@
...
-from ...compat import BaseModel
-from ...ty import FewshotExample, Self
+from ...compat import BaseModel, Self
+from ...ty import FewshotExample, TaskContraT
 
 
-class SpanExample(FewshotExample, abc.ABC):
+class SpanExample(FewshotExample[TaskContraT], abc.ABC, Generic[TaskContraT]):
+    """Example for span tasks not using CoT.
+    Note: this should be SpanTaskContraT instead of TaskContraT, but this would entail a circular import.
```
Comment (svlandeg) [inline spacy_llm/tasks/span/examples.py:12]: Isn't there any other solution? A circular import typically means we should reorganize what is in which module...

File: spacy_llm/tasks/span/task.py [hunk 3]
```diff
@@ -62,7 +63,7 @@ def __init__(
         self._description = description
 
         if self._prompt_examples:
-            self._prompt_examples = list(self._check_label_consistency(self))
+            self._prompt_examples = list(check_label_consistency(self))
```
Comment (svlandeg) [inline spacy_llm/tasks/span/task.py:66]: I think it's fine to use the original code, refering to the method in `self`, as this was defined on L61. (imagine we'd change the definition on L61 in a future refactor, then we might miss that we're still using the "old" one here)

File: spacy_llm/ty.py [hunk 4]
```diff
@@ -111,14 +102,24 @@ def parse_responses(
...
+TaskContraT = TypeVar("TaskContraT", bound=LLMTask, contravariant=True)
+
+
+class FewshotExample(GenericModel, abc.ABC, Generic[TaskContraT]):
+    @classmethod
+    @abc.abstractmethod
+    def generate(cls, example: Example, task: TaskContraT) -> Self:
+        """Create a fewshot example from a spaCy example.
+        example (Example): spaCy example.
+        task (_TaskContraT): Task for which to generate examples.
```
Comment (svlandeg) [inline spacy_llm/ty.py:114]:
```suggestion
        task (TaskContraT): Task for which to generate examples.
```

**Review by svlandeg — COMMENTED** [review]
Mostly looks good!

As this removes 3.6 support, I'd target `develop` instead of `main` (though it might end up being the same release)

File: spacy_llm/tasks/span/examples.py (hunk 2 above)
Comment (rmitsch) [reply spacy_llm/tasks/span/examples.py:12]: I agree, but in this case: 
- `SpanExample` should use `FewshotExample` parametrized with `SpanTaskContraT`: `FewshotExample[SpanTaskContraT]`.
- `SpanTaskContraT` requires `SpanTask` in its definition.
- `SpanTask` requires `SpanExample`.

This is a dependency loop, i. e. doing this in multiple files will (without local imports) always lead to circular import errors. The one alternative here would be to drop everything in `task.py` - also not nice :man_shrugging:

File: spacy_llm/tasks/rel/util.py (hunk 1 above)
Comment (rmitsch) [reply spacy_llm/tasks/rel/util.py]: Outdated, removed it.

File: spacy_llm/tasks/span/task.py (hunk 3 above)
Comment (rmitsch) [reply spacy_llm/tasks/span/task.py]: Done in https://github.com/explosion/spacy-llm/pull/306/commits/19d3660926260ae1f159c782add4caeaf83476db.

**Pull request merged**